cmd/chainbridge/deploycontracts.go

```diff
@@ -0,0 +1,271 @@
+	log.Info("Deploying Contracts")
+
+	port := ctx.String(PortFlag.Name)
+	if port == "" {
+		port = "8545"
+	}
+
+	validators := ctx.Int(NumValidatorsFlag.Name)
+	if validators == 0 {
+		validators = 2
+	}
+
+	validatorThreshold := ctx.Int(ValidatorThresholdFlag.Name)
+	if validatorThreshold == 0 {
+		validatorThreshold = 2
+	}
+
+	depositThreshold := ctx.Int(DepositThresholdFlag.Name)
+	if depositThreshold == 0 {
+		depositThreshold = 2
+	}
+
+	minCount := ctx.Int(MinCountFlag.Name)
+	if minCount == 0 {
+		minCount = 10
+	}
+
+	deployPK := ctx.String(DeployPKFlag.Name)
+	if deployPK == "" {
+		deployPK = DEPLOYER_PRIV_KEY
+	}
+
+	recieverAddr, emitterAddr, simpleEmitterAddr, bridgeAssetAddr, err := deployContractsLocal(deployPK, port, validators, big.NewInt(int64(validatorThreshold)), big.NewInt(int64(depositThreshold)), uint8(minCount))
+	if err != nil {
+		return err
+	}
+
+	log.Info("Reciever Contract Deployed at: " + recieverAddr.Hex())
+	log.Info("Emitter Contract Deployed at: " + emitterAddr.Hex())
+	log.Info("Simple Emitter Contract Deployed at: " + simpleEmitterAddr.Hex())
+	log.Info("Bridge Asset Contract Deployed at: " + bridgeAssetAddr.Hex())
+
+	return nil
+}
+
+func deployContractsLocal(deployPK string, port string, validators int, validatorThreshold *big.Int, depositThreshold *big.Int, minCount uint8) (common.Address, common.Address, common.Address, common.Address, error) {
+
+	client, auth, deployAddress, validatorAddresses, err := accountSetUp(port, validators, deployPK)
+	if err != nil {
+		log.Error(err.Error())
+		return ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, err
+	}
+
+	recieverAddr, err := deployReceiver(auth, client, validatorAddresses, depositThreshold, validatorThreshold, deployAddress)
+	if err != nil {
+		return ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, err
+	}
+
+	emitterAddr, err := deployEmitter(auth, client, deployAddress)
+	if err != nil {
+		return ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, err
+	}
+
+	simpleEmitterAddr, err := deploySimpleEmitter(auth, client, deployAddress)
+	if err != nil {
+		return ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, err
+	}
+
+	bridgeAssetAddr, err := deployBridgeAsset(auth, client, minCount, deployAddress)
+	if err != nil {
+		return ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, err
+	}
+
+	return recieverAddr, emitterAddr, simpleEmitterAddr, bridgeAssetAddr, nil
+
+}
+
+func createValidatorSlice(valAddr []string, numValidators int) []common.Address {
+
+	validatorAddresses := make([]common.Address, numValidators)
+
+	for i := 0; i < numValidators; i++ {
+		validatorAddresses[i] = common.HexToAddress(valAddr[i])
+
+	}
+
+	return validatorAddresses
+}
+
+func updateNonce(auth *bind.TransactOpts, client *ethclient.Client, deployAddress common.Address) (*bind.TransactOpts, error) {
+
+	newNonce, err := client.PendingNonceAt(context.Background(), deployAddress)
+	if err != nil {
+		log.Error(err.Error())
```

Comment:
You're getting into some ugly logging here. This is logged and returned to a method that will also log and return the same error. This is always a bit of a challenge to do concisely.

In this case it's probably best to return errors instead of logging them, since all errors will be returned to the CLI and eventually logged. It may be helpful to add context to errors, best to use `fmt.Errorf()` to extend the returned error. There are a few places this could be updated.

Reply:
I'll try to only log the error at the base level, thanks david